internal/handlers/bank_send_restriction_test.go

```diff
@@ -0,0 +1,117 @@
+	addr2beforeBalance := app.BankKeeper.GetAllBalances(ctx, addr2).String()
+	assert.Equal(tt, "1000000000stake", addr2beforeBalance, "addr2beforeBalance")
+
+	// send withdraw for coins
+	withdrawMsg := markertypes.NewMsgWithdrawRequest(addr1, addr1, nrMarkerDenom,
+		sdk.NewCoins(sdk.NewInt64Coin(nrMarkerDenom, 1000)))
+	ConstructAndSendTx(tt, *app, ctx, acct1, priv, withdrawMsg, abci.CodeTypeOK, "")
+	// Check both account balances before we begin.
+	addr1afterBalance := app.BankKeeper.GetAllBalances(ctx, addr1).String()
+	assert.Equal(tt, "1000nonrestrictedmarker,999850000stake", addr1afterBalance, "addr1afterBalance")
+
+	// send withdraw for coins
+	withdrawMsg = markertypes.NewMsgWithdrawRequest(addr1, addr1, rMarkerDenom,
+		sdk.NewCoins(sdk.NewInt64Coin(rMarkerDenom, 1000)))
+	ConstructAndSendTx(tt, *app, ctx, acct1, priv, withdrawMsg, abci.CodeTypeOK, "")
+	addr1afterBalance = app.BankKeeper.GetAllBalances(ctx, addr1).String()
+	assert.Equal(tt, "1000nonrestrictedmarker,1000restrictedmarker,999700000stake", addr1afterBalance, "addr1afterBalance")
+
+	// send restricted marker from account with transfer rights and no required attributes, expect success
+	sendRMarker := banktypes.NewMsgSend(addr1, addr2, sdk.NewCoins(sdk.NewInt64Coin(rMarkerDenom, 100)))
+	ConstructAndSendTx(tt, *app, ctx, acct1, priv, sendRMarker, abci.CodeTypeOK, "")
+	addr1afterBalance = app.BankKeeper.GetAllBalances(ctx, addr1).String()
+	assert.Equal(tt, "1000nonrestrictedmarker,900restrictedmarker,999550000stake", addr1afterBalance, "addr1afterBalance")
+	addr2afterBalance := app.BankKeeper.GetAllBalances(ctx, addr2).String()
+	assert.Equal(tt, "100restrictedmarker,1000000000stake", addr2afterBalance, "addr2beforeBalance")
+
+	// send non restricted marker, expect success
+	sendRMarker = banktypes.NewMsgSend(addr1, addr2, sdk.NewCoins(sdk.NewInt64Coin(nrMarkerDenom, 100)))
+	ConstructAndSendTx(tt, *app, ctx, acct1, priv, sendRMarker, abci.CodeTypeOK, "")
+	addr1afterBalance = app.BankKeeper.GetAllBalances(ctx, addr1).String()
+	assert.Equal(tt, "900nonrestrictedmarker,900restrictedmarker,999400000stake", addr1afterBalance, "addr1afterBalance")
+	addr2afterBalance = app.BankKeeper.GetAllBalances(ctx, addr2).String()
+	assert.Equal(tt, "100nonrestrictedmarker,100restrictedmarker,1000000000stake", addr2afterBalance, "addr2beforeBalance")
+
+	sendRMarker = banktypes.NewMsgSend(addr2, addr1, sdk.NewCoins(sdk.NewInt64Coin(nrMarkerDenom, 50)))
+	ConstructAndSendTx(tt, *app, ctx, acct2, priv2, sendRMarker, abci.CodeTypeOK, "")
+	addr1afterBalance = app.BankKeeper.GetAllBalances(ctx, addr1).String()
+	assert.Equal(tt, "950nonrestrictedmarker,900restrictedmarker,999400000stake", addr1afterBalance, "addr1afterBalance")
+	addr2afterBalance = app.BankKeeper.GetAllBalances(ctx, addr2).String()
+	assert.Equal(tt, "50nonrestrictedmarker,100restrictedmarker,999850000stake", addr2afterBalance, "addr2beforeBalance")
+
+	stopIfFailed(tt)
```

Comment:
First, you don't need this `stopIfFailed(tt)` call at the end since it doesn't do anything that the unit test won't already do at the end of a test.

Second, It'd be good to have a couple other test cases that demonstrate some failed `MsgSend`s.
1. On a restricted coin without required attributes where the sender doesn't have `TRANSFER` permission.
2. On a restricted coin with required attributes from a sender that has `TRANSFER` permission, but the receiver doesn't have the required attributes.
3. On a restricted coin with required attributes from a sender that does not have `TRANSFER` permission, but the receiver DOES have the required attributes.

Also some failed `MsgTransfer`s:
1. On a restricted coin with required attributes using an admin that has `TRANSFER` permission, but the receiver doesn't have the required attributes.
2. On a restricted coin with required attributes using an admin that does not have `TRANSFER` permission, but the receiver DOES have the required attributes.